src/runtime/propensity.js

```diff
@@ -69,27 +76,44 @@ export class Propensity {
 
   /** @override */
   sendEvent(userEvent) {
-    if (!Object.values(PropensityApi.Event).includes(userEvent.name)) {
-      throw new Error('Invalid user event provided');
-    }
-    if (userEvent.data && !isObject(userEvent.data)) {
-      throw new Error('Event data must be an Object');
+    const analyticsEvent = propensityEventToAnalyticsEvent(userEvent.name);
+    let data = null;
+    if (!isEnumValue(PropensityApi.Event, userEvent.name)
+        || !analyticsEvent) {
+      throw new Error('Invalid user event provided(' + userEvent.name + ')');
     }
-    let paramString = null;
-    if (userEvent.active != null) {
-      if (!userEvent.data) {
-        const jsonData = Object.create(null);
-        jsonData['is_active'] = userEvent.active;
-        userEvent.data = /** @type {JsonObject} */(jsonData);
+
+    if (userEvent.data) {
+      if (!isObject(userEvent.data)) {
+        throw new Error('Event data must be an Object(' + userEvent.data + ')');
       } else {
-        Object.assign(userEvent.data, {'is_active': userEvent.active});
+        data = {};
+        Object.assign(data, userEvent.data);
       }
     }
-    if (userEvent.data) {
-      paramString = JSON.stringify(userEvent.data);
+
+    if (isBoolean(userEvent.active)) {
+      if (!data) {
+        data = {};
+      }
+      Object.assign(data, {'is_active': userEvent.active});
+    } else if (userEvent.active != null) {
```

Comment:
Now that we are not enforcing userEvent.active to be null and not undefined in case the caller has no information about the event being from a user action or not, we need to update your Event Manager code to not fail validation of the event and corresponding tests. You can either make that change first as a different PR or roll that into this one.

Reply:
I looked into this. Your point is valid. We need to send something, undefined won't do. However, I wanted to find a way to handle it that is consistent with the rest of our code based. I found several instances of '|| null' to cover cases where undefined could be passed into an API for the field of an object defined using typedef.

Whenever a typedef requires something to be set and won't accept undefined, we OR the incoming object with a null. In this case it would look like this.

    this.eventManager_.logEvent({
    this.propensityServer_.sendEvent(userEvent.name, paramString);	      eventType: analyticsEvent,
      eventOriginator: EventOriginator.PROPENSITY_CLIENT,
      isFromUserAction: userEvent.active || null,
      additionalParameters: data,
    });

I still think we don't have to have a hard check to ensure its not undefined because the code already doesn't allow it. So != is sufficient IMHO. But to your point, to make it comply, following the code pattern in the rest of the code and ORing with null should address your concerns. 
 Let me know your thoughts.

Reply:
given that we have a new way to handle undefined, can we update the message on 101 to be Active field must be boolean. 
Because if its not boolean, both null or undefined are okay. we handle it in line 107.

Reply:
I'm still quite convinced that only test code can ever pass in undefined because the compiler will reject anything that allows undefined.  The reason it works in the tests is that the compiler seems to be less strict for them.  Real code won't be able to do it (except perhaps through Propensity where the publishers might not be doing strict compilation checking).  

I'd also strongly prefer not allowing it to be either undefined or null and treating them as the same, but it isn't actually all that important so I've just gone ahead and made that change.

Reply:
I didn't see the previous 2 comments.  I don't think the || null is necessary because Propensity.Event doesn't support not having active.

Reply:
Believe we have a resolution to this.